### PR TITLE
fix getElementsByTagName()

### LINKED ISSUE #11596: Properly implement the getElementsByTagName method on Document and Element
These two methods don't pass the tests that use non-HTML elements.

See the two tests "Element in non-HTML namespace, prefix, uppercase name" and "Element in non-HTML namespace, prefix, lowercase name".

We should think about how to implement efficiently the step 2 of [the spec](https://dom.spec.whatwg.org/#concept-getelementsbytagname).

See methods `Document::GetElementsByTagName`, `Element::GetElementsByTagName`, `HTMLCollection::by_tag_name` and `HTMLCollection::by_atomic_tag_name`.

These two last methods should be documented and probably renamed to something a bit more descriptive.

Files:
- components/script/dom/document.rs
- components/script/dom/element.rs
- components/script/dom/htmlcollection.rs
- tests/wpt/metadata/dom/nodes/Document-getElementsByTagName.html.ini
- tests/wpt/metadata/dom/nodes/Element-getElementsByTagName.html.ini
- tests/wpt/web-platform-tests/dom/nodes/Document-Element-getElementsByTagName.js
- tests/wpt/web-platform-tests/dom/nodes/Document-getElementsByTagName.html
- tests/wpt/web-platform-tests/dom/nodes/Element-getElementsByTagName.html

### PR DESCRIPTION
**Pull request by mskrzypkows**
Improved implementation of getElementsByTagName() in Document, properly pass 3 cases of https://dom.spec.whatwg.org/#concept-getelementsbytagname

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix  #11596 (github issue number if applicable).

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14047)

### CODE REVIEW

**Comment by mskrzypkows** [comment]
continuation of #12844

**Comment by mskrzypkows** [comment]
@nox r?

File: components/script/dom/htmlcollection.rs [hunk 1]
```diff
@@ -113,39 +112,53 @@ impl HTMLCollection {
...
+        if qualified_name == atom!("*") {
+            #[derive(JSTraceable, HeapSizeOf)]
+            struct AllFilter;
+            impl CollectionFilter for AllFilter {
+                fn filter(&self, _elem: &Element, _root: &Node) -> bool {
+                    true
+                }
+            }
+            return HTMLCollection::create(window, root, box AllFilter);
+        }
```
Comment (mskrzypkows) [inline components/script/dom/htmlcollection.rs]: OK, so should I add additional checking?

**Comment by nox** [comment]
@bors-servo r+

Thanks for keeping up with it!

**Comment by nox** [comment]
@mskrzypkows Almost here, just need to remove that failure expectation.

**Pull request merged**